### PR TITLE
feat: price-feeder telemetry

### PR DESCRIPTION
**Pull request by adamewozniak**
## Description

Adds support for prometheus telemetry to our price feeder. Get via `curl localhost:7171/api/v1/metrics`.
Example response [here](https://pastebin.com/L2MypxdD)
closes: #378

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: price-feeder/oracle/oracle.go [hunk 1]
```diff
@@ -98,12 +99,19 @@ func (o *Oracle) Start(ctx context.Context) error {
...
+			startTime := time.Now()
+
 			if err := o.tick(); err != nil {
+				telemetry.IncrCounter(1, "failure", "tick")
 				o.logger.Err(err).Msg("oracle tick failed")
 			}
 
 			o.lastPriceSyncTS = time.Now()
 
+			telemetry.MeasureSince(startTime, "runtime", "tick")
```
Comment (adamewozniak) [inline price-feeder/oracle/oracle.go:112]: I don't think we should include `time.Sleep` as a part of this metric, although it's technically a part of the `tick`, would make the data more difficult to evaluate

File: price-feeder/telemetry/metrics.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+// Prometheus or JSON if in-memory.
+func (m *Metrics) Gather(format string) (GatherResponse, error) {
+	switch format {
+	case FormatPrometheus:
+		return m.gatherPrometheus()
+
+	case FormatText:
+		return m.gatherGeneric()
+
+	case FormatDefault:
```
Comment (alexanderbez) [inline price-feeder/telemetry/metrics.go:130]: I think we can combine these? `case FormatText, FormatDefault:`. Give that a shot.

File: price-feeder/telemetry/metrics.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	metricsConf.EnableHostnameLabel = cfg.EnableHostnameLabel
+
+	memSink := metrics.NewInmemSink(10*time.Second, time.Minute)
+	metrics.DefaultInmemSignal(memSink)
+
+	m := &Metrics{memSink: memSink}
+	fanout := metrics.FanoutSink{memSink}
+
+	if cfg.PrometheusRetentionTime > 0 {
+		m.prometheusEnabled = true
```
Comment (alexanderbez) [inline price-feeder/telemetry/metrics.go:99]: So I know this was ported from the SDK (which I wrote actually), but realizing all this time later, I think there is a slightly cleaner way for operators. I would suggest we do the following:

1. Remove the `PrometheusRetentionTime` config.
2. Add a `Type` config instead, whose value must be `generic` or `prometheus`.
3. Hardcode `PrometheusRetentionTime` to something like 120 or whatever.

File: price-feeder/router/v1/response.go [hunk 4]
```diff
@@ -24,3 +27,22 @@ type (
...
+func newErrorResponse(code int, err string) errorResponse {
+	return errorResponse{Code: code, Error: err}
+}
+
+// writeErrorResponse prepares and writes a HTTP error
+// given a status code and an error message.
+func writeErrorResponse(w http.ResponseWriter, status int, err string) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(status)
+	_, _ = w.Write(legacy.Cdc.MustMarshalJSON(newErrorResponse(0, err)))
```
Comment (alexanderbez) [inline price-feeder/router/v1/response.go:47]: No need for this dependency. Let's just use `json` directly.

```suggestion
  bz, _ := json.Marshal(newErrorResponse(0, err))
	_, _ = w.Write(bz)
```

File: price-feeder/oracle/client/client.go [hunk 5]
```diff
@@ -144,6 +145,7 @@ func (oc OracleClient) BroadcastTx(nextBlockHeight int64, timeoutHeight int64, m
 
 		resp, err := BroadcastTx(clientCtx, factory, msgs...)
 		if resp != nil && resp.Code != 0 {
+			telemetry.IncrCounter(1, "failure", "tx")
```
Comment (alexanderbez) [inline price-feeder/oracle/client/client.go:148]: Do you think we should add code here?

File: price-feeder/oracle/client/client.go [hunk 6]
```diff
@@ -144,6 +145,7 @@ func (oc OracleClient) BroadcastTx(nextBlockHeight int64, timeoutHeight int64, m
...
 			err = fmt.Errorf("invalid response code from tx")
```
Comment (alexanderbez) [inline price-feeder/oracle/client/client.go:149]: While we're at it, I forgot to mention in the other PR, let's add a `code` field to the error message.

File: price-feeder/config/config.go [hunk 7]
```diff
@@ -45,12 +46,13 @@ var (
...
-		Account       Account        `toml:"account" validate:"required,gt=0,dive,required"`
-		Keyring       Keyring        `toml:"keyring" validate:"required,gt=0,dive,required"`
-		RPC           RPC            `toml:"rpc" validate:"required,gt=0,dive,required"`
-		GasAdjustment float64        `toml:"gas_adjustment" validate:"required"`
+		Server        Server           `toml:"server"`
+		CurrencyPairs []CurrencyPair   `toml:"currency_pairs" validate:"required,gt=0,dive,required"`
+		Account       Account          `toml:"account" validate:"required,gt=0,dive,required"`
+		Keyring       Keyring          `toml:"keyring" validate:"required,gt=0,dive,required"`
+		RPC           RPC              `toml:"rpc" validate:"required,gt=0,dive,required"`
+		Telemetry     telemetry.Config `toml:"telemetry"`
```
Comment (alexanderbez) [inline price-feeder/config/config.go:54]: I would bet that telemetry config should probably exist here as `Telemetry`?

**Review by alexanderbez — COMMENTED** [review]
Amazing work 👏 🚀

File: price-feeder/oracle/client/client.go (hunk 6 above)
Comment (adamewozniak) [reply price-feeder/oracle/client/client.go]: Updated these lines to : 
```go
telemetry.IncrCounter(1, "failure", "tx", "code")
err = fmt.Errorf("invalid response code from tx: %d", resp.Code)
```

**Review by alexanderbez — APPROVED** [review]

**Pull request merged**